File: static/js/identity.js
/**
 * identity.js — Identity aggregation, votes, weekly review. Port of core/identity.py.
 *
 * Depends on: periods.js, stats.js, habitDB (db.js) for persistence.
 * NO dependency on Flask or SQLite.
 */

const identity = (() => {

  function _plural(n, one, many) {
    return n === 1 ? one : many;
  }

  async function identityVotes(ident, nowDt) {
    const allHabits = await habitDB.listHabits();
    const identHabits = allHabits.filter(h => h.identity_id === ident.id && h.active);
    let events = [];
    for (const h of identHabits) {
      const hEvents = await habitDB.listEvents(h.id);
      events = events.concat(hEvents);
    }
    events.sort((a, b) => (a.occurred_at || "").localeCompare(b.occurred_at || ""));
    const [ws] = periods.periodThisWeek(nowDt);
    const [ms] = periods.periodThisMonth(nowDt);
    const startDate = "1970-01-01";
    return {
      this_week: stats.countIn(events, ws, nowDt),
      this_month: stats.countIn(events, ms, nowDt),
      total: events.length,
      week_average: Math.round(stats.weeklyAverage(events, startDate, nowDt) * 10) / 10,
      week_average_last4: Math.round(stats.weeklyAverage(events, startDate, nowDt, 4) * 10) / 10,
      active_days: stats.activeDays(events),
      best_week: stats.bestWeek(events, startDate, nowDt),
      trajectory: stats.weeklyCounts(events, startDate, nowDt, 12),
      trend: stats.trend(events, startDate, nowDt),
    };
  }

  function identityEvidence(votes) {
    const out = [];
    const total = votes.total;
    if (total === 0) {
      return ["Todavía no hay acciones registradas para esta identidad. Cada pequeña acción será un voto."];
    }
    out.push("Has realizado " + total + " " + _plural(total, "acción", "acciones") + " relacionadas con esta identidad.");
    out.push("Esta semana has actuado como esta persona " + votes.this_week + " " + _plural(votes.this_week, "vez", "veces") + ".");
    if (votes.week_average_last4 > 0) {
      out.push("En las últimas 4 semanas has mantenido una media de " + votes.week_average_last4 + " acciones por semana.");
    }
    const d = votes.trend.direction;
    if (d === "up") out.push("Tu frecuencia está aumentando: estás construyendo más evidencia cada semana.");
    else if (d === "down") out.push("El ritmo ha bajado un poco. Un periodo bajo no borra la evidencia ya construida.");
    return out;
  }

  async function identitySummary(ident, nowDt) {
    const votes = await identityVotes(ident, nowDt);
    const allHabits = await habitDB.listHabits();
    const habitCount = allHabits.filter(h => h.identity_id === ident.id && h.active).length;
    return { ...ident, votes, habit_count: habitCount };
  }

  async function habitsForIdentity(identityId) {
    const all = await habitDB.listHabits();
    return all
      .filter(h => h.identity_id === identityId && h.active)
      .sort((a, b) => (a.created_at || "").localeCompare(b.created_at || ""));
  }

  function frictionStatus(habit, events, nowDt) {
    const { period, target } = stats.periodTarget(habit);
    if (target <= 0) return null;
    const startDate = habit.start_date || "1970-01-01";
    if (period === "day") {
      const buckets = periods.dayBuckets(startDate, nowDt, 10);
      const counts = buckets.map(b => stats.countIn(events, b.start, b.end));
      const completed = (counts.length && counts[counts.length - 1] < target) ? counts.slice(0, -1) : counts;
      const recent = completed.slice(-4);
      const below = recent.filter(c => c < target).length;
      if (recent.length >= 3 && recent[recent.length - 1] < target && below >= 3)
        return { difficult: true, reason: "several_days", missed: below };
      return null;
    }
    const buckets = period === "month"
      ? periods.monthBuckets(startDate, nowDt)
      : periods.weekBuckets(startDate, nowDt);
    const counts = buckets.map(b => stats.countIn(events, b.start, b.end));
    const completed = (counts.length && counts[counts.length - 1] < target) ? counts.slice(0, -1) : counts;
    const recent = completed.slice(-3);
    const below = recent.filter(c => c < target).length;
    if (recent.length >= 2 && recent[recent.length - 1] < target && below >= 2)
      return { difficult: true, reason: "repeated_misses", missed: below };
    return null;
  }

  const REVIEW_DAY = 7;

  async function reviewSchedule(nowDt) {
    const raw = await habitDB.getSetting("review_sunday_time");
    if (!raw) return { enabled: false, day: REVIEW_DAY, time: null, open: false, next_at: null };
    let hour, minute;
    try {
      const parts = raw.split(":").map(Number);
      hour = parts[0]; minute = parts[1];
      if (hour < 0 || hour > 23 || minute < 0 || minute > 59) throw new Error();
    } catch (_) {
      return { enabled: false, day: REVIEW_DAY, time: null, open: false, next_at: null };
    }
    const isoDay = nowDt.getDay() === 0 ? 7 : nowDt.getDay();
    const isSunday = isoDay === REVIEW_DAY;
    const openNow = isSunday && (nowDt.getHours() > hour || (nowDt.getHours() === hour && nowDt.getMinutes() >= minute));
    const nextAt = periods.nextWeekdayAt(nowDt, REVIEW_DAY, hour, minute);
    const pad2 = n => String(n).padStart(2, "0");
    return {
      enabled: true,
      day: REVIEW_DAY,
      time: pad2(hour) + ":" + pad2(minute),
      open: openNow,
      next_at: nextAt.getFullYear() + "-" + pad2(nextAt.getMonth() + 1) + "-" + pad2(nextAt.getDate()) + "T" + pad2(nextAt.getHours()) + ":" + pad2(nextAt.getMinutes()) + ":" + pad2(nextAt.getSeconds()),
    };
  }

  async function weeklyReview(nowDt, habitSummaryFn) {
    const habits = (await habitDB.listHabits()).filter(h => h.active);
    habits.sort((a, b) => (a.created_at || "").localeCompare(b.created_at || ""));
    const [ws] = periods.startOfWeek(nowDt);
    const we = periods.addDays(ws, 7);
    const weekStart = ws.getFullYear() + "-" + String(ws.getMonth() + 1).padStart(2, "0") + "-" + String(ws.getDate()).padStart(2, "0");
    const wsIso = weekStart + "T00:00";

    const rows = [];
    for (const h of habits) {
      const events = (await habitDB.listEvents(h.id)).sort((a, b) => (a.occurred_at || "").localeCompare(b.occurred_at || ""));
      rows.push({ habit: h, events, summary: habitSummaryFn(h) });
    }

    const working = [];
    const difficult = [];
    for (const r of rows) {
      const g = r.summary.goal || {};
      if (g.achieved && (g.period_count || 0) > 0) {
        working.push({ ...r.summary, week_count: g.period_count });
      } else if (frictionStatus(r.habit, r.events, nowDt)) {
        difficult.push({ ...r.summary, week_count: g.this_week || 0 });
      }
    }
    working.sort((a, b) => b.week_count - a.week_count);

    const identities = (await habitDB.listIdentities()).filter(i => i.active !== false);
    identities.sort((a, b) => (a.name || "").localeCompare(b.name || ""));
    const identResult = [];
    for (const ident of identities) {
      const allHabitsForIdent = (await habitDB.listHabits()).filter(h => h.identity_id === ident.id && h.active);
      let thisWeekCount = 0;
      for (const h of allHabitsForIdent) {
        const evts = await habitDB.listEvents(h.id);
        thisWeekCount += stats.countIn(evts, ws, nowDt);
      }
      if (thisWeekCount > 0) identResult.push({ ...ident, this_week: thisWeekCount });
    }
    identResult.sort((a, b) => b.this_week - a.this_week);

    const cueCounts = {};
    const habitCues = [];
    for (const r of rows) {
      const cue = (r.habit.cue || "").trim();
      const weekCount = stats.countIn(r.events, ws, nowDt);
      if (cue && weekCount > 0) {
        cueCounts[cue] = (cueCounts[cue] || 0) + weekCount;
        habitCues.push([r.habit.name, cue, weekCount]);
      }
    }
    const patterns = [];
    if (Object.keys(cueCounts).length) {
      let topCue = "", topN = 0;
      for (const [cue, n] of Object.entries(cueCounts)) { if (n > topN) { topN = n; topCue = cue; } }
      const names = habitCues.filter(([_, c]) => c === topCue).map(([n]) => n).join(", ");
      patterns.push("La señal más frecuente esta semana ha sido \u00ab" + topCue + "\u00bb (" + topN + " registros).");
    }
    if (working.length) {
      const top = working[0];
      patterns.push("\u00ab" + top.name + "\u00bb ha sido el hábito más consistente esta semana (" + top.week_count + " acciones).");
    }
    if (difficult.length) {
      patterns.push("\u00ab" + difficult[0].name + "\u00bb parece estar costando. Puedes revisar el sistema para hacerlo más fácil.");
    }

    const reviewAnswer = await habitDB.getReviewAnswer(weekStart);
    const answer = reviewAnswer ? reviewAnswer.answer : null;

    return {
      week_start: weekStart,
      working,
      difficult,
      identities: identResult,
      patterns,
      question_answer: answer,
      total_actions: rows.reduce((sum, r) => sum + (r.summary.goal || {}).this_week || 0, 0),
    };
  }

  return {
    _plural,
    identityVotes,
    identityEvidence,
    identitySummary,
    habitsForIdentity,
    frictionStatus,
    REVIEW_DAY,
    reviewSchedule,
    weeklyReview,
  };
})();

File: core/identity.py
"""Identity aggregation, votes, weekly review and patterns.

"Votes" are simply events: every realization of a habit linked to an identity
is one vote for that identity. They are never turned into scores or levels.
"""
from datetime import timedelta

from . import db, periods, stats


def identity_votes(identity, conn, now_dt):
    """Aggregate votes (events) for an identity across its habits."""
    events = conn.execute(
        """SELECT e.* FROM events e
           JOIN habits h ON h.id = e.habit_id
           WHERE h.identity_id = ? AND h.active = 1 AND e.occurred_at >= '1970-01-01T00:00'
           ORDER BY e.occurred_at ASC, e.id ASC""",
        (identity["id"],),
    ).fetchall()
    events = [dict(r) for r in events]

    ws, _we = periods.period_this_week(now_dt)
    ms, _me = periods.period_this_month(now_dt)
    start_date = "1970-01-01"

    this_week = stats.count_in(events, ws, now_dt)
    this_month = stats.count_in(events, ms, now_dt)
    total = len(events)
    week_average = stats.weekly_average(events, start_date, now_dt)
    week_average_last4 = stats.weekly_average(events, start_date, now_dt, weeks=4)
    active_days = stats.active_days(events)
    best_week = stats.best_week(events, start_date, now_dt)
    trajectory = stats.weekly_counts(events, start_date, now_dt, weeks=12)
    tr = stats.trend(events, start_date, now_dt)

    return {
        "this_week": this_week,
        "this_month": this_month,
        "total": total,
        "week_average": round(week_average, 1),
        "week_average_last4": round(week_average_last4, 1),
        "active_days": active_days,
        "best_week": best_week,
        "trajectory": trajectory,
        "trend": tr,
    }


def identity_evidence(votes, now_dt):
    """Descriptive, evidence-based sentences. Never a score."""
    out = []
    total = votes["total"]
    if total == 0:
        return ["Todavía no hay acciones registradas para esta identidad. Cada pequeña acción será un voto."]
    out.append(
        f"Has realizado {total} {_plural(total, 'acción', 'acciones')} relacionadas con esta identidad."
    )
    out.append(
        f"Esta semana has actuado como esta persona {votes['this_week']} "
        f"{_plural(votes['this_week'], 'vez', 'veces')}."
    )
    if votes["week_average_last4"] > 0:
        out.append(
            f"En las últimas 4 semanas has mantenido una media de "
            f"{votes['week_average_last4']} acciones por semana."
        )
    d = votes["trend"]["direction"]
    if d == "up":
        out.append("Tu frecuencia está aumentando: estás construyendo más evidencia cada semana.")
    elif d == "down":
        out.append("El ritmo ha bajado un poco. Un periodo bajo no borra la evidencia ya construida.")
    return out


def identity_summary(identity, conn, now_dt):
    """Identity plus vote aggregates and habit count."""
    votes = identity_votes(identity, conn, now_dt)
    return {
        **identity,
        "votes": votes,
        "habit_count": conn.execute(
            "SELECT COUNT(*) AS n FROM habits WHERE identity_id = ? AND active = 1", (identity["id"],)
        ).fetchone()["n"],
    }


def habits_for_identity(conn, identity_id):
    rows = conn.execute(
        "SELECT * FROM habits WHERE identity_id = ? AND active = 1 ORDER BY created_at ASC",
        (identity_id,),
    ).fetchall()
    return [dict(r) for r in rows]


# ---------------------------------------------------------------------------
# Friction detection
# ---------------------------------------------------------------------------

def friction_status(habit, events, now_dt):
    """Detect habits that seem difficult. Never guilt — just a signal to
    review the system. Returns None or a dict with a reason."""
    period, target = stats.period_target(habit)
    if target <= 0:
        return None
    start_date = habit.get("start_date") or "1970-01-01"

    if period == "day":
        buckets = periods.day_buckets(start_date, now_dt, days=10)
        counts = [stats.count_in(events, b[1], b[2]) for b in buckets]
        completed = counts[:-1] if counts and counts[-1] < target else counts
        recent = completed[-4:]
        below = sum(1 for c in recent if c < target)
        if len(recent) >= 3 and recent[-1] < target and below >= 3:
            return {"difficult": True, "reason": "several_days", "missed": below}
        return None

    buckets = periods.month_buckets(start_date, now_dt) if period == "month" else periods.week_buckets(start_date, now_dt)
    counts = [stats.count_in(events, b[1], b[2]) for b in buckets]
    completed = counts[:-1] if counts and counts[-1] < target else counts
    recent = completed[-3:]
    below = sum(1 for c in recent if c < target)
    if len(recent) >= 2 and recent[-1] < target and below >= 2:
        return {"difficult": True, "reason": "repeated_misses", "missed": below}
    return None


# ---------------------------------------------------------------------------
# Weekly review
# ---------------------------------------------------------------------------

REVIEW_DAY = 7  # Sunday (ISO)


def review_schedule(conn, now_dt):
    """Weekly review moment configured by the user.

    The user picks an hour for Sunday. The review page only opens on Sunday
    from that hour until the day ends; the rest of the week a countdown shows
    the time until the next opening.
    """
    raw = db.get_setting(conn, "review_sunday_time")
    if not raw:
        return {"enabled": False, "day": REVIEW_DAY, "time": None,
                "open": False, "next_at": None}
    try:
        hour, minute = (int(part) for part in raw.split(":"))
        if not (0 <= hour <= 23 and 0 <= minute <= 59):
            raise ValueError
    except (TypeError, ValueError):
        return {"enabled": False, "day": REVIEW_DAY, "time": None,
                "open": False, "next_at": None}
    is_sunday = now_dt.isoweekday() == REVIEW_DAY
    open_now = is_sunday and (now_dt.hour, now_dt.minute) >= (hour, minute)
    next_at = periods.next_weekday_at(now_dt, REVIEW_DAY, hour, minute)
    return {
        "enabled": True,
        "day": REVIEW_DAY,
        "time": f"{hour:02d}:{minute:02d}",
        "open": open_now,
        "next_at": next_at.strftime("%Y-%m-%dT%H:%M:%S"),
    }


def weekly_review(conn, now_dt, habit_summary_fn):
    """Build the weekly review payload.

    habit_summary_fn(habit) -> dict with at least: id, name, icon, color,
    goal, identity info.
    """
    habits = [dict(r) for r in conn.execute(
        "SELECT * FROM habits WHERE active = 1 ORDER BY created_at ASC"
    ).fetchall()]
    ws = periods.start_of_week(now_dt)
    we = ws + timedelta(days=7)
    week_start = ws.strftime("%Y-%m-%d")

    rows = []
    for h in habits:
        events = conn.execute(
            "SELECT * FROM events WHERE habit_id = ? ORDER BY occurred_at ASC, id ASC", (h["id"],)
        ).fetchall()
        events = [dict(r) for r in events]
        rows.append({"habit": h, "events": events, "summary": habit_summary_fn(h)})

    # What worked: habits meeting their objective this week, best first.
    working = []
    difficult = []
    for r in rows:
        g = r["summary"].get("goal") or {}
        if g.get("achieved") and g.get("period_count", 0) > 0:
            working.append({**r["summary"], "week_count": g["period_count"]})
        elif friction_status(r["habit"], r["events"], now_dt):
            difficult.append({**r["summary"], "week_count": g.get("this_week", 0)})
    working.sort(key=lambda x: x["week_count"], reverse=True)

    # Identities reinforced this week.
    identities = []
    for identity in conn.execute("SELECT * FROM identities WHERE active = 1 ORDER BY name ASC").fetchall():
        identity = dict(identity)
        this_week = conn.execute(
            """SELECT COUNT(*) AS n FROM events e JOIN habits h ON h.id = e.habit_id
               WHERE h.identity_id = ? AND h.active = 1 AND e.occurred_at >= ? AND e.occurred_at < ?""",
            (identity["id"], ws.strftime("%Y-%m-%dT%H:%M"), we.strftime("%Y-%m-%dT%H:%M")),
        ).fetchone()["n"]
        if this_week > 0:
            identities.append({**identity, "this_week": this_week})
    identities.sort(key=lambda x: x["this_week"], reverse=True)

    # Patterns: which cues appear most among habits with registrations this week.
    cue_counts = {}
    habit_cues = []
    for r in rows:
        cue = (r["habit"].get("cue") or "").strip()
        week_count = stats.count_in(r["events"], ws, now_dt)
        if cue and week_count > 0:
            cue_counts[cue] = cue_counts.get(cue, 0) + week_count
            habit_cues.append((r["habit"]["name"], cue, week_count))
    patterns = []
    if cue_counts:
        top_cue, top_n = max(cue_counts.items(), key=lambda kv: kv[1])
        names = ", ".join(n for n, c, wc in habit_cues if c == top_cue)
        patterns.append(
            f"La señal más frecuente esta semana ha sido «{top_cue}» ({top_n} registros)."
        )
    if working:
        top = working[0]
        patterns.append(f"«{top['name']}» ha sido el hábito más consistente esta semana ({top['week_count']} acciones).")
    if difficult:
        patterns.append(f"«{difficult[0]['name']}» parece estar costando. Puedes revisar el sistema para hacerlo más fácil.")

    answer_row = conn.execute("SELECT * FROM reviews WHERE week_start = ?", (week_start,)).fetchone()
    answer = dict(answer_row)["answer"] if answer_row else None

    return {
        "week_start": week_start,
        "working": working,
        "difficult": difficult,
        "identities": identities,
        "patterns": patterns,
        "question_answer": answer,
        "total_actions": sum(r["summary"].get("goal", {}).get("this_week", 0) for r in rows),
    }


def _plural(n, one, many):
    return one if n == 1 else many